### PR TITLE
Adding Publish Code Changes for vstest node implementation

### PR DESCRIPTION
**Pull request by prawalagarwal**
Problem: The vstest task needed the implementation for publishing code changes
Fix: Added the required changes
Testing: Manual

### CODE REVIEW

File: Tasks/VsTest/vstest.ts [hunk 1]
```diff
@@ -219,56 +241,70 @@ function executeVstest(testResultsDirectory: string, parallelRunSettingsFile: st
...
+                            .fail(function (code) {
+                                defer.resolve(code);
+                            });
+                    });
             })
-            .fail(function (err) {
+            .fail(function (err){
                 tl.error(err);
-                tl.warning(tl.loc('ErrorWhileCreatingResponseFile'));
+                tl.warning(tl.loc('ErrorWhilePublishingCodeChanges'));
```
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: Add this new error message to task.json

File: Tasks/VsTest/vstest.ts [hunk 2]
```diff
@@ -187,6 +187,28 @@ function generateResponseFile(): Q.Promise<string> {
     return defer.promise;
 }
 
+function publishCodeChanges(): Q.Promise<string> {
```
Comment (acesiddhu) [inline Tasks/VsTest/vstest.ts]: i thnk we shuld not publish code changes if already done. shuld it be in tool or at task layer? what do u think. Also how will we do that?

File: Tasks/VsTest/vstest.ts [hunk 3]
```diff
@@ -187,6 +187,28 @@ function generateResponseFile(): Q.Promise<string> {
...
+    var defer = Q.defer<string>();
+    var selectortool = tl.createToolRunner(tiaSelectorTool);
+    selectortool.arg("PublishCodeChanges");
+    selectortool.arg("/TfsTeamProjectCollection:" + tl.getVariable("System.TeamFoundationCollectionUri"));
+    selectortool.arg("/ProjectId:" + tl.getVariable("System.TeamProject"));
+    selectortool.arg("/newdropPath:" + tl.getVariable("Build.SourcesDirectory"));
+    selectortool.arg("/Definitionid:" + tl.getVariable("System.DefinitionId"));
+    selectortool.arg("/buildid:" + tl.getVariable("Build.BuildId"));
+    selectortool.arg("/token:" + tl.getVariable("System.AccessToken"));
```
Comment (acesiddhu) [inline Tasks/VsTest/vstest.ts]: is this variable available for task by default?

Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: Availability of tiaSelectorTool will be taken care by isTiaAllowed func which will check for existence tiaSelectorTool before invoking any tia related functions

File: Tasks/VsTest/vstest.ts [hunk 4]
```diff
@@ -187,6 +187,28 @@ function generateResponseFile(): Q.Promise<string> {
...
+
+    selectortool.exec()
```
Comment (acesiddhu) [inline Tasks/VsTest/vstest.ts]: we r not printing output logs of tool. check how vstest exec is done

File: Tasks/VsTest/vstest.ts [hunk 5]
```diff
@@ -187,6 +187,28 @@ function generateResponseFile(): Q.Promise<string> {
     return defer.promise;
 }
 
+function publishCodeChanges(): Q.Promise<string> {
+    var defer = Q.defer<string>();
```
Comment (acesiddhu) [inline Tasks/VsTest/vstest.ts]: can we add perf counter here?

Comment (acesiddhu) [inline Tasks/VsTest/vstest.ts]: as well as abve method

---
In reply to: [69462821](https://github.com/Microsoft/vsts-tasks/pull/2006#discussion_r69462821) [](ancestors = 69462821)

File: Tasks/VsTest/vstest.ts (hunk 4 above)
Comment (prawalagarwal) [inline Tasks/VsTest/vstest.ts]: The documentation of exec code says this:
"Exec a tool.
     * Output will be streamed to the live console."
Exec will take care of the output

File: Tasks/VsTest/vstest.ts [hunk 6]
```diff
@@ -166,6 +166,9 @@ function updateResponseFile(argsArray: string[], responseFile: string): Q.Promis
 
 
 function generateResponseFile(): Q.Promise<string> {
+    var startTime = perfMeasure();
```
Comment (acesiddhu) [inline Tasks/VsTest/vstest.ts]: validate if this gets published to server

**Comment by acesiddhu** [comment]
:shipit:

File: definitions/vsts-task-lib.d.ts [hunk 7]
```diff
@@ -248,6 +248,26 @@ declare module 'vsts-task-lib/task' {
...
+	export function getEndpointDataParameter(id: string, key: string, optional: boolean): string;
+	/**
+	 * Gets the endpoint authorization scheme for a service endpoint
+	 * If the endpoint authorization scheme is not set and is not optional, the task will fail with an error message. Execution will halt.
+	 *
+	 * @param id name of the service endpoint
+	 * @param optional whether the endpoint authorization scheme is optional
+	 * @returns {string} value of the endpoint authorization scheme
+	 */
+	export function getEndpointAuthorizationScheme(id: string, optional: boolean): string;
```
Comment (acesiddhu) [inline definitions/vsts-task-lib.d.ts]: looks fine. just make sure end to end is working fine.

**Pull request merged**